base_layer/wallet/src/tasks/wallet_recovery.rs

```diff
@@ -26,49 +26,328 @@ use crate::{
+            (&self.peer_seed_public_keys[self.peer_seed_public_keys.len() - 1]).clone()
+        };
+        self.rpc_retries = 1;
+        loop {
+            // Allow PRC connections to be retried N times before using up a retry
+            if self.rpc_retries > RPC_RETRIES_LIMIT {
+                self.rpc_retries = 1;
+                self.connection_retries += 1;
+            }
+            if self.connection_retries > self.connection_retries_limit {
+                return Err(WalletError::WalletRecoveryError(format!(
+                    "Could not connect to base node within the specified number of retries ({})",
+                    self.connection_retries_limit
+                )));
+            }
+
+            let mut reconnect_to_base_node = true;
+            if let Some(connection) = self.base_node_connection.clone() {
+                reconnect_to_base_node = !connection.is_connected();
+                trace!(
+                    target: LOG_TARGET,
+                    "Base node {} is connected {}",
+                    self.base_node_public_key.clone(),
+                    connection.is_connected(),
+                );
+            }
+
+            if reconnect_to_base_node {
+                // Select next base node in list to try and connect to, wrapping around
+                let mut new_base_node_public_key = last_base_node_public_key.clone();
+                if select_new_base_node {
+                    for i in 0..(self.peer_seed_public_keys.len()) {
+                        if self.peer_seed_public_keys[i] == last_base_node_public_key.clone() {
+                            if i != self.peer_seed_public_keys.len() - 1 {
+                                new_base_node_public_key = (&self.peer_seed_public_keys[i + 1]).clone();
+                            } else {
+                                new_base_node_public_key = (&self.peer_seed_public_keys[0]).clone();
+                            }
+                            last_base_node_public_key = new_base_node_public_key.clone();
+                            break;
+                        }
+                    }
+                }
+
+                // Attempt new base node connection
+                debug!(
+                    target: LOG_TARGET,
+                    "Trying to connect to {} (retries {} of {})",
+                    new_base_node_public_key,
+                    self.connection_retries,
+                    self.connection_retries_limit
+                );
+                if select_new_base_node && self.print_to_console {
+                    println!(
+                        "Trying to connect to {} (retries {} of {})",
+                        new_base_node_public_key.clone(),
+                        self.connection_retries,
+                        self.connection_retries_limit
+                    );
+                }
+                let base_node_node_id = NodeId::from_key(&new_base_node_public_key.clone()).map_err(|e| e)?;
```

Comment:
Use the infallible `from_public_key` method 
```suggestion
                let base_node_node_id = NodeId::from_public_key(&new_base_node_public_key);
```